File: 06-gcd-of-three-numbers/python/main.py
#!/usr/bin/python3

def init() -> tuple[int, int, int]:
    """ Вводимо початкові дані. """

    a = int(input("Введіть a: "))
    b = int(input("Введіть b: "))
    c = int(input("Введіть c: "))

    return a, b, c


def gcd(a: int, b: int, c: int = 0) -> int:
    """ Знаходить найбільше спільне кратне ДВОХ або ТРЬОХ чисел. """

    def __gcd(a: int, b: int) -> int:
        """ Знаходить найбільше спільне кратне ДВОХ чисел.

        Використовує алгоритм Евкліда. """

        if a < b:
            a, b = b, a

        while b != 0:
            a %= b
            a, b = b, a

        return abs(a)

    return __gcd(__gcd(a, b), c)


def browse(a: int, b: int, c: int, result: int) -> None:
    """ Виводимо результат обчислень. """

    print(f"НСД{a, b, c} = {result}")


if __name__ == "__main__":
    a, b, c = init()
    result = gcd(a, b, c)
    browse(a, b, c, result)

File: 06-gcd-of-three-numbers/cpp/main.cpp
#include <iostream>

using namespace std;

void init(int &, int &, int &);   // введення даних
int gcd(int, int, int);           // обчислення найбільшого спільного дільника
template <typename T> void exchange(T &, T &); // обмін двох змінних
void browse(int, int, int, int);  // вивдення результатів

int main() {
    int a, b, c,
        result;

    init(a, b, c);
    result = gcd(a, b, c);
    browse(a, b, c, result);

    return 0;
}

void init(int &a, int &b, int &c) {
    cout << "Введіть а: "; cin >> a;
    cout << "Введіть b: "; cin >> b;
    cout << "Введіть c: "; cin >> c;
}

// Знаходить найбільше спільне кратне ТРЬОХ чисел.
// Використувує власитвість НСД(a, b, c) = НСД(НСД(a, b), c)
inline int gcd(int a, int b, int c) {
    int gcd(int, int);

    return gcd(gcd(a, b), c);
}

template <typename T>
void exchange(T &a, T &b) {
    T temp;
    temp = a;
    a = b;
    b = temp;
}

// Знаходить найбільше спільне кратне ДВОХ чисел.
// Використовує алгоритм Евкліда.
int gcd(int a, int b) {
    if (a < b)
        exchange(a, b);

    while (b != 0) {
        a %= b;
        exchange(a, b);
    }

    return abs(a);
}


void browse(int a, int b, int c, int result) {
    cout << "НСД(" << a << ", " << b << ", " << c << ") = " << result << endl;
}
